### PR TITLE
feat: add xrpl

### PR DESCRIPTION
**Pull request by Polybius93**
This PR adds XRPL network as option for the minting/redeeming flow.

### CODE REVIEW

File: config.testnet.json [hunk 1]
```diff
@@ -1,13 +1,16 @@
...
   "enabledEthereumNetworkIDs": ["421614", "84532", "11155111"],
+  "enabledRippleNetworkIDs": ["1"],
   "bitcoinNetwork": "testnet",
   "bitcoinNetworkIndex": 1,
   "bitcoinNetworkPreFix": "tb1",
   "bitcoinBlockchainURL": "https://testnet.dlc.link/electrs",
   "bitcoinBlockchainExplorerURL": "https://mempool.space/testnet",
   "bitcoinBlockchainFeeEstimateURL": "https://mempool.space/testnet/api/v1/fees/recommended",
+  "rippleIssuerAddress": "ra9epzthPkNXykgfadCwu8D7mtajj8DVCP",
+  "xrplWebsocket": "wss://testnet.xrpl-labs.com/",
```
Comment (sosaucily) [inline config.testnet.json:13]: This seems like it might be a problem when it points to our internal node. 

This is a public repo, we shouldn't be exposing our IPs and URLs. Shouldn't this be either a netlify config value, and perhaps in a netlify serverless function?

File: src/app/components/mint-unmint/components/burn-transaction-screen/burn-transaction-screen.tsx [hunk 2]
```diff
@@ -34,48 +31,40 @@ export function BurnTokenTransactionForm({
...
-  const { chainId } = useAccount();
+  const { ethereumNetworkConfiguration } = useContext(EthereumNetworkConfigurationContext);
 
   const signer = useEthersSigner();
 
   const { unmintStep } = useSelector((state: RootState) => state.mintunmint);
   const [isSubmitting, setIsSubmitting] = useState(false);
 
-  const currentVault = allVaults.find(vault => vault.uuid === unmintStep[1]);
+  const currentVault = unmintStep[2];
```
Comment (sosaucily) [inline src/app/components/mint-unmint/components/burn-transaction-screen/burn-transaction-screen.tsx:47]: Generally it would be better if you're not using hardcoded numbers here, but you can set up some global consts for the steps and use those variables?

File: src/app/components/mint-unmint/components/deposit-transaction-screen/deposit-transaction-screen.tsx [hunk 3]
```diff
@@ -48,8 +48,10 @@ export function DepositTransactionScreen({
 
     try {
       setIsSubmitting(true);
-      const currentRisk = await fetchUserEthereumAddressRiskLevel();
-      if (currentRisk === 'High') throw new Error('Risk Level is too high');
+      if (networkType === 'evm') {
+        const currentRisk = await fetchUserEthereumAddressRiskLevel();
+        if (currentRisk === 'High') throw new Error('Risk Level is too high');
```
Comment (sosaucily) [inline src/app/components/mint-unmint/components/deposit-transaction-screen/deposit-transaction-screen.tsx:53]: Is this the message we print when chainalysis returns a high risk level? If so, ask me for some better text. If not, what's this

File: src/app/components/mint-unmint/components/walkthrough/components/walkthrough-blockchain-tag.tsx [hunk 4]
```diff
@@ -1,18 +1,22 @@
...
     text: 'ON ETHEREUM',
   },
   bitcoin: {
     logo: '/images/logos/bitcoin-logo.svg',
     text: 'ON BITCOIN',
   },
+  xrpl: {
+    logo: '/images/logos/xrp-logo.svg',
+    text: 'ON XRPL',
+  },
```
Comment (sosaucily) [inline src/app/components/mint-unmint/components/walkthrough/components/walkthrough-blockchain-tag.tsx:15]: if the network types were in a global list, like ALL_NETWORKS, then you can loop through them in various situations, and handle each case, and then if there are any missing, you can throw an error or better yet, Typescript will even through an error. So you can know at compile time that there is somewhere in the code that you forgot to handle the new network type. 

so maybe, if it took networks of type Bitcoin | EVM | XRPL and then you add a new type, typescript would blow up. Does that make sense? I wonder if that would be helpful

File: src/app/components/mint-unmint/components/walkthrough/walkthrough.tsx [hunk 5]
```diff
@@ -22,20 +27,28 @@ export function Walkthrough({ flow, currentStep }: WalkthroughProps): React.JSX.
...
-                  color={'accent.lightBlue.01'}
-                  href="https://metamask.io/"
-                  isExternal
-                  textDecoration={'underline'}
-                >
-                  Ethereum Wallet
-                </Link>
-                .
-              </Text>
+              {networkType === 'evm' ? (
```
Comment (sosaucily) [inline src/app/components/mint-unmint/components/walkthrough/walkthrough.tsx:32]: for example here, you have a switch to handle these types, but if you miss one you wouldn't know it until you actually run the code and select the network of that missing type

File: src/app/components/modals/select-wallet-modal/select-wallet-modal.tsx [hunk 6]
```diff
@@ -1,73 +1,177 @@
...
+
+  const [selectedNetworkID, setSelectedNetworkID] = useState<
+    EthereumNetworkID | RippleNetworkID | undefined
+  >(undefined);
+
   const [selectedWagmiConnectorID, setSelectedWagmiConnectorID] = useState<string | undefined>(
     undefined
   );
 
+  const networkTypes = ['evm', 'xrpl'];
```
Comment (sosaucily) [inline src/app/components/modals/select-wallet-modal/select-wallet-modal.tsx:65]: here as well, you're just defining this list all over the place in the codebase. why not define it once globally, and ideally leverage type checking

File: src/app/components/select-network-button/select-network-button.tsx [hunk 7]
```diff
@@ -1,39 +1,67 @@
...
+import { EthereumNetwork, EthereumNetworkID } from 'dlc-btc-lib/models';
 
 interface SelectNetworkButtonProps {
-  handleChangeNetwork: (ethereumNetwork: Chain) => void;
-  ethereumNetworks: readonly [Chain, ...Chain[]];
-  selectedEthereumNetwork?: Chain;
+  handleChangeNetwork: (networkID: EthereumNetworkID | RippleNetworkID) => void;
+  ethereumNetworkIDs: EthereumNetworkID[];
+  rippleNetworkIDs: RippleNetworkID[];
+  selectedNetworkType?: 'evm' | 'xrpl';
```
Comment (sosaucily) [inline src/app/components/select-network-button/select-network-button.tsx:13]: here too

File: src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx [hunk 8]
```diff
@@ -49,6 +49,7 @@ export function TransactionFormSubmitButtonGroup({
   handleCancelButtonClick,
   isSubmitting,
 }: TransactionFormSubmitButtonGroupProps): React.JSX.Element {
+  console.log('userEthereumAddressRiskLevel', userEthereumAddressRiskLevel);
```
Comment (sosaucily) [inline src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx:52]: console.log, i guess the linter would be warning you about this

Comment (sosaucily) [reply src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx:52]: if you want to keep it, can you exclude it?

File: src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/transaction-screen.transaction-form.tsx [hunk 9]
```diff
@@ -141,22 +140,31 @@ export function VaultTransactionForm({
...
-            ),
+            assetAmount: validateFormAmount(decimalValue.toNumber(), flow, depositLimit, vault),
           },
         };
       },
     },
   });
 
+  function parseAssetAmount(assetAmount: string): Decimal {
+    const isValidNumber = /^-?\d*\.?\d*$/.test(assetAmount);
```
Comment (sosaucily) [inline src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/transaction-screen.transaction-form.tsx:156]: could add this to a utils file or something, so you can use it in other places when needed rather than forgetting it's here and duplicating your code

Comment (sosaucily) [reply src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/transaction-screen.transaction-form.tsx:156]: or maybe Ramda has a function for that? probably not

File: src/app/components/vault/components/vault.detaills/vault.details.tsx [hunk 10]
```diff
@@ -28,30 +31,40 @@ export function VaultDetails({
...
+    if (handleClose) {
+      handleClose();
+    }
   }
 
   function handleWithdrawClick() {
     navigate('/mint-withdraw');
     if (vaultTotalLockedValue === vaultTotalMintedValue) {
-      dispatch(mintUnmintActions.setUnmintStep([0, vaultUUID]));
+      dispatch(mintUnmintActions.setUnmintStep([0, vaultUUID, vault]));
```
Comment (sosaucily) [inline src/app/components/vault/components/vault.detaills/vault.details.tsx:53]: steps as integers

File: src/app/hooks/use-confirmation-checker.ts [hunk 11]
```diff
@@ -34,7 +34,12 @@ export function useConfirmationChecker(): [string, number][] {
     try {
       const bitcoinTransactionBlockHeight = await fetchBitcoinTransactionBlockHeight(vault);
 
-      return blockHeight - bitcoinTransactionBlockHeight;
+      const difference = blockHeight - bitcoinTransactionBlockHeight;
+      if (difference < 0) {
+        return 0;
+      } else {
+        return difference;
+      }
```
Comment (sosaucily) [inline src/app/hooks/use-confirmation-checker.ts:42]:
```ts
return Math. max(difference, 0)
```

File: src/app/hooks/use-connected.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+  const [isConnected, setIsConnected] = useState<boolean>(false);
+
+  const { networkType } = useContext(NetworkConfigurationContext);
+
+  const { isConnected: isEthereumWalletConnected } = useAccount();
+  const { xrpWalletContextState } = useContext(XRPWalletContext);
+
+  useEffect(() => {
+    switch (networkType) {
+      case 'evm':
```
Comment (sosaucily) [inline src/app/hooks/use-connected.ts:21]: types!

File: src/app/hooks/use-psbt.ts [hunk 13]
```diff
@@ -52,24 +67,55 @@ export function usePSBT(): UsePSBTReturnType {
...
+    } else {
+      throw new Error('Network Type is not setup');
+    }
+  };
+
+  const getAttestorChainID = (): string => {
+    if (networkType === 'evm') {
+      return ethereumAttestorChainID;
+    } else if (networkType === 'xrpl') {
+      return rippleAttestorChainID;
```
Comment (sosaucily) [inline src/app/hooks/use-psbt.ts:102]: good place to iterate through a typed list i think

File: src/app/hooks/use-xrpl-ledger.ts [hunk 14]
```diff
@@ -0,0 +1,165 @@
+import { useContext, useState } from 'react';
```
Comment (sosaucily) [inline src/app/hooks/use-xrpl-ledger.ts:1]: is this file copied into our repo, or did you write from scratch?

File: src/app/providers/network-configuration.provider.tsx [hunk 15]
```diff
@@ -0,0 +1,27 @@
+import React, { createContext, useState } from 'react';
+
+import { HasChildren } from '@models/has-children';
+
+interface NetworkConfigurationContext {
+  networkType: 'evm' | 'xrpl';
+  setNetworkType: (networkType: 'evm' | 'xrpl') => void;
```
Comment (sosaucily) [inline src/app/providers/network-configuration.provider.tsx:7]: types

**Review by sosaucily — COMMENTED** [review]
Quite good! mostly these are just improvement ideas, except perhaps for the configuration file recommendation. 

Also, there are quite a lot of lint errors. Can you add a lint-exception for those? or delete them

File: config.testnet.json (hunk 1 above)
Comment (Polybius93) [reply config.testnet.json:13]: I was not planning to use our internal node for the website, public works well with the website, since the requests are minimal. 

I would rather not pass entire handlers (which are using this endpoint) back and forth from the app to netlify.

File: src/app/components/mint-unmint/components/burn-transaction-screen/burn-transaction-screen.tsx (hunk 2 above)
Comment (Polybius93) [reply src/app/components/mint-unmint/components/burn-transaction-screen/burn-transaction-screen.tsx:47]: So this is not the step's index, but unmintStep is a triplet and we are using the third value here.

File: src/app/components/vault/components/vault.detaills/vault.details.tsx (hunk 10 above)
Comment (Polybius93) [reply src/app/components/vault/components/vault.detaills/vault.details.tsx:53]: Alright, I can make this an enum.

File: src/app/hooks/use-xrpl-ledger.ts (hunk 14 above)
Comment (Polybius93) [reply src/app/hooks/use-xrpl-ledger.ts:1]: I wrote this, why are you assuming it was copied over?

File: config.testnet.json [hunk 16]
```diff
@@ -1,13 +1,16 @@
...
+  "coordinatorURL": "https://testnet-ripple.dlc.link/attestor-1",
   "enabledEthereumNetworkIDs": ["421614", "84532", "11155111"],
+  "enabledRippleNetworkIDs": ["1"],
   "bitcoinNetwork": "testnet",
   "bitcoinNetworkIndex": 1,
   "bitcoinNetworkPreFix": "tb1",
   "bitcoinBlockchainURL": "https://testnet.dlc.link/electrs",
   "bitcoinBlockchainExplorerURL": "https://mempool.space/testnet",
   "bitcoinBlockchainFeeEstimateURL": "https://mempool.space/testnet/api/v1/fees/recommended",
+  "rippleIssuerAddress": "ra9epzthPkNXykgfadCwu8D7mtajj8DVCP",
```
Comment (kskulikova) [inline config.testnet.json:12]: We have a new one now I believe ("ra3oyRVfy4yD4NJPrVcewvDtisZ3FhkcYL")

File: src/app/components/mint-unmint/components/burn-transaction-screen/burn-transaction-screen.tsx (hunk 2 above)
Comment (sosaucily) [reply src/app/components/mint-unmint/components/burn-transaction-screen/burn-transaction-screen.tsx]: I'm sure there's a way to not use integers directly here, but rather friendly named vars

Comment (Polybius93) [reply src/app/components/mint-unmint/components/burn-transaction-screen/burn-transaction-screen.tsx]: Yeah, I modified it!

File: src/app/components/header/header.tsx [hunk 17]
```diff
@@ -22,14 +29,13 @@ export function Header(): React.JSX.Element {
   const isMobile = useBreakpointValue({ base: true, md: false });
 
   useEffect(() => {
-    if (isConnected && !chain) {
+    if (networkType === NetworkType.EVM && isConnected && !ethereumNetwork) {
```
Comment (sosaucily) [inline src/app/components/header/header.tsx:32]: do you need to check both !ethereumNetwork and networkType === networkType.evm? seems redundant?

File: src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx [hunk 18]
```diff
@@ -49,6 +49,9 @@ export function TransactionFormSubmitButtonGroup({
   handleCancelButtonClick,
   isSubmitting,
 }: TransactionFormSubmitButtonGroupProps): React.JSX.Element {
+  const isButtonDisabledByRiskLevel =
+    flow === 'burn' && currentStep === 0 && userEthereumAddressRiskLevel === 'High';
```
Comment (sosaucily) [inline src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx:53]: I thought i might have left that comment before, but what about naming the steps, rather than remembering the numbers for the different steps? Seems it would make it much easier to change the steps in the future, and prevent mistakes.

File: src/app/components/vaults-list/components/vaults-list-group-container.tsx [hunk 19]
```diff
@@ -29,7 +35,7 @@ export function VaultsListGroupContainer({
             {['Pending'].includes(label) && <Spinner color={'accent.lightBlue.01'} size={'md'} />}
             <Text color={'white'}>{label}</Text>
           </HStack>
-          {label === 'Minted dlcBTC' && (
+          {label === 'Minted dlcBTC' && networkType === NetworkType.EVM && (
```
Comment (sosaucily) [inline src/app/components/vaults-list/components/vaults-list-group-container.tsx:38]: comparing against a label seems really dangerous. If the label changes (which it might do for lots of reasons) then this logic will fail.

File: src/app/hooks/use-active-tabs.ts [hunk 20]
```diff
@@ -17,20 +22,32 @@ export function useActiveTabs(): UseActiveTabsReturnType {
...
 
   const { data: isActiveTabs } = useQuery({
-    queryKey: ['activeTabs', chain, address],
+    queryKey: ['activeTabs', chain, address, networkType, xrpWalletContextState],
     queryFn: shouldActivateTabs,
-    enabled: !isEthereumNetworkConfigurationLoading,
+    enabled:
+      networkType === NetworkType.EVM
+        ? !isEthereumNetworkConfigurationLoading
+        : !isRippleNetworkConfigurationLoading,
```
Comment (sosaucily) [inline src/app/hooks/use-active-tabs.ts:50]: there is lots and lots of conditional logic all around this codebase now. I'm guessing this is something you also want to clean up. Doesn't have to be in this PR, but the more conditional logic around, the harder it will be to make changes in the future.

File: src/app/hooks/use-confirmation-checker.ts [hunk 21]
```diff
@@ -34,7 +34,8 @@ export function useConfirmationChecker(): [string, number][] {
     try {
       const bitcoinTransactionBlockHeight = await fetchBitcoinTransactionBlockHeight(vault);
 
-      return blockHeight - bitcoinTransactionBlockHeight;
+      const difference = blockHeight - bitcoinTransactionBlockHeight;
+      return Math.max(difference, 0);
```
Comment (sosaucily) [inline src/app/hooks/use-confirmation-checker.ts:38]: Looks like you are fixing something about the confirmation checker?  does this need to be in this PR?

File: src/shared/utils.ts [hunk 22]
```diff
@@ -51,6 +51,19 @@ export function formatEvent(event: DetailedEvent): FormattedEvent {
...
+  const isValidNumber = /^-?\d*\.?\d*$/.test(assetAmount);
+
+  if (isValidNumber) {
+    try {
+      return new Decimal(assetAmount);
+    } catch {
+      return new Decimal(0);
+    }
+  }
+  return new Decimal(0);
```
Comment (sosaucily) [inline src/shared/utils.ts:64]: this also could be implemented in another PR? Or is it something just for this PR?

File: src/app/hooks/use-confirmation-checker.ts (hunk 21 above)
Comment (Polybius93) [reply src/app/hooks/use-confirmation-checker.ts]: Done, moved to a different PR!

File: src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx (hunk 18 above)
Comment (Polybius93) [reply src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx:53]: Yes, forgot to use the Enum here!

File: src/app/components/vaults-list/components/vaults-list-group-container.tsx (hunk 19 above)
Comment (Polybius93) [reply src/app/components/vaults-list/components/vaults-list-group-container.tsx:38]: Okay, added additional argument which represents the State of the vaults.

File: src/app/hooks/use-active-tabs.ts (hunk 20 above)
Comment (Polybius93) [reply src/app/hooks/use-active-tabs.ts:50]: Yes, it would be nice to refactor some flow related conditions in a future PR.

**Review by sosaucily — APPROVED** [review]

File: src/app/components/mint-unmint/components/deposit-transaction-screen/deposit-transaction-screen.tsx (hunk 3 above)
Comment (Polybius93) [reply src/app/components/mint-unmint/components/deposit-transaction-screen/deposit-transaction-screen.tsx:56]: At this point, if the risk level is High, we would have already seen some text message warning the user, this is just some extra safety step. But can be refactored in a future PR.

File: src/app/components/mint-unmint/components/walkthrough/components/walkthrough-blockchain-tag.tsx (hunk 4 above)
Comment (Polybius93) [reply src/app/components/mint-unmint/components/walkthrough/components/walkthrough-blockchain-tag.tsx:17]: Added NetworkType enum, which includes EVM, BTC, XRPL. Replaced string type with NetworkType enum arguments.

File: src/app/components/mint-unmint/components/walkthrough/walkthrough.tsx (hunk 5 above)
Comment (Polybius93) [reply src/app/components/mint-unmint/components/walkthrough/walkthrough.tsx]: Well, I wouldn't want to have different cases for every network type. But I replaced this string type with the NetworkType enum here too.

File: src/app/components/modals/select-wallet-modal/select-wallet-modal.tsx (hunk 6 above)
Comment (Polybius93) [reply src/app/components/modals/select-wallet-modal/select-wallet-modal.tsx]: Replaced with NetworkType Enum

File: src/app/components/select-network-button/select-network-button.tsx (hunk 7 above)
Comment (Polybius93) [reply src/app/components/select-network-button/select-network-button.tsx]: Replaced with NetworkType Enum

File: src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx (hunk 8 above)
Comment (Polybius93) [reply src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/components/transaction-screen.transaction-form.submit-button-group.tsx]: Removed!

File: src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/transaction-screen.transaction-form.tsx (hunk 9 above)
Comment (Polybius93) [reply src/app/components/transaction-screen/transaction-screen.transaction-form/components/transaction-screen.transaction-form/transaction-screen.transaction-form.tsx]: Moved to utils file

File: src/app/hooks/use-connected.ts (hunk 12 above)
Comment (Polybius93) [reply src/app/hooks/use-connected.ts]: Replaced with NetworkType Enum

File: src/app/hooks/use-psbt.ts (hunk 13 above)
Comment (Polybius93) [reply src/app/hooks/use-psbt.ts]: Replaced with NetworkType Enum

File: src/app/hooks/use-xrpl-ledger.ts (hunk 14 above)
Comment (sosaucily) [reply src/app/hooks/use-xrpl-ledger.ts:1]: Just curious
